pkg/agent/route/route_linux.go

```diff
@@ -0,0 +1,568 @@
+// It's idempotent and can be safely called on every startup.
+func (c *Client) Initialize(nodeConfig *config.NodeConfig) error {
+	c.nodeConfig = nodeConfig
+
+	// Ensure required ipsets exist.
+	if err := ipset.CreateIPSet(antreaPodIPSet, ipset.HashNet); err != nil {
+		return fmt.Errorf("failed to creeate ipset: %v", err)
+	}
+
+	// Sets up the iptables infrastructure required to route packets in host network.
+	if err := c.initIPTables(); err != nil {
+		return fmt.Errorf("failed to initialize iptables: %v", err)
+	}
+
+	// Sets up the IP routes and IP rule required to route packets in host network.
+	if err := c.initIPRoutes(); err != nil {
+		return fmt.Errorf("failed to initialize ip routes: %v", err)
+	}
+
+	// send_redirects must be disabled because packets from hostGateway are
+	// routed back to it. Otherwise redirect packets will be sent to source
+	// Pods.
+	// send_redirects for the interface will be enabled if at least one of
+	// conf/{all,interface}/send_redirects is set to TRUE, so "all" and the
+	// interface must be disabled together.
+	// See https://www.kernel.org/doc/Documentation/networking/ip-sysctl.txt.
+	if err := disableICMPSendRedirects("all"); err != nil {
+		return err
+	}
+	if err := disableICMPSendRedirects(c.hostGateway); err != nil {
+		return err
+	}
+	return nil
+}
+
+// initIPTables ensure that the iptables infrastructure we use is set up.
+// It's idempotent and can safely be called on every startup.
+func (c *Client) initIPTables() error {
+	// Create the antrea managed chains and link them to built-in chains.
+	// We cannot use iptables-restore for these jump rules because there
+	// are non antrea managed rules in built-in chains.
+	jumpRules := []struct{ table, srcChain, dstChain, comment string }{
+		{iptables.FilterTable, iptables.ForwardChain, AntreaForwardChain, "Antrea: jump to Antrea forwarding rules"},
+		{iptables.NATTable, iptables.PostRoutingChain, AntreaPostRoutingChain, "Antrea: jump to Antrea postrouting rules"},
+		{iptables.MangleTable, iptables.PreRoutingChain, antreaMangleChain, "Antrea: jump to Antrea mangle rules"},
+		{iptables.RawTable, iptables.PreRoutingChain, antreaRawChain, "Antrea: jump to Antrea raw rules"},
+	}
+	for _, rule := range jumpRules {
+		if err := c.ipt.EnsureChain(rule.table, rule.dstChain); err != nil {
+			return err
+		}
+		ruleSpec := []string{"-j", rule.dstChain, "-m", "comment", "--comment", rule.comment}
+		if err := c.ipt.EnsureRule(rule.table, rule.srcChain, ruleSpec); err != nil {
+			return err
+		}
+	}
+
+	// Create required rules in the antrea chains.
+	// Use iptables-restore as it flushes the involved chains and creates the desired rules
+	// with a single call, instead of string matching to clean up stale rules.
+	iptablesData := bytes.NewBuffer(nil)
+	writeLine(iptablesData, "*mangle")
+	writeLine(iptablesData, iptables.MakeChainLine(antreaMangleChain))
+	writeLine(iptablesData, []string{
```

Comment:
They are designed to replace the original dynamical added rules:
https://github.com/vmware-tanzu/antrea/blob/8687e60fec1727f28dfec30a464aeda88dda6901/pkg/agent/iptables/iptables.go#L177-L180.

In encap, `-i gw0 ! -o gw0` can identify external traffic and will be masqueraded;
In noEncap and passthrough, there were dynamic rules inserted before the above rule to skip masquerading as pod-to-pod can go out from gw0; but the logic is not very straightforward and may slow down forwarding in scale and needs update on node events:
```
-i gw0 ! -o gw0 -d <REMOTE PODCIDR A> -s <LOCAL PODCIDR> -j ACCEPT
-i gw0 ! -o gw0 -d <REMOTE PODCIDR B> -s <LOCAL PODCIDR> -j ACCEPT
-i gw0 ! -o gw0 -d <REMOTE PODCIDR C> -s <LOCAL PODCIDR> -j ACCEPT
-i gw0 ! -o gw0 -j mark --mark xxxx
```
After this patch, it will just be 3 rules (the second rule will be present anyway):
```
-i gw0 -m set --match-set ANTREA-POD-IP dst -m comment --comment "Antrea: mark pod to pod packets" -j MARK --set-xmark 0x200/0x200
-d <SERVICE CIDR> -i gw0 -m comment --comment "Antrea: mark pod to service packets" -j MARK --set-xmark 0x800/0x800
-i gw0 -m mark --mark 0x0/0xa00 -m comment --comment "Antrea: mark pod to external packets" -j MARK --set-xmark 0x400/0x400
```
Added explanation to code too.